tools/data_rewriter.py

```diff
@@ -0,0 +1,168 @@
+    'Illinois': 'IL',
+    'Indiana': 'IN',
+    'Iowa': 'IA',
+    'Kansas': 'KS',
+    'Kentucky': 'KY',
+    'Louisiana': 'LA',
+    'Maine': 'ME',
+    'Maryland': 'MD',
+    'Massachusetts': 'MA',
+    'Michigan': 'MI',
+    'Minnesota': 'MN',
+    'Mississippi': 'MS',
+    'Missouri': 'MO',
+    'Montana': 'MT',
+    'Nebraska': 'NE',
+    'Nevada': 'NV',
+    'New Hampshire': 'NH',
+    'New Jersey': 'NJ',
+    'New Mexico': 'NM',
+    'New York': 'NY',
+    'North Carolina': 'NC',
+    'North Dakota': 'ND',
+    'Northern Mariana Islands':'MP',
+    'Ohio': 'OH',
+    'Oklahoma': 'OK',
+    'Oregon': 'OR',
+    'Pennsylvania': 'PA',
+    'Puerto Rico': 'PR',
+    'Rhode Island': 'RI',
+    'South Carolina': 'SC',
+    'South Dakota': 'SD',
+    'Tennessee': 'TN',
+    'Texas': 'TX',
+    'Utah': 'UT',
+    'Vermont': 'VT',
+    'Virgin Islands': 'VI',
+    'Virginia': 'VA',
+    'Washington': 'WA',
+    'Washington DC': 'DC',
+    'West Virginia': 'WV',
+    'Wisconsin': 'WI',
+    'Wyoming': 'WY',
+    'Unknown Location': unknown_location_acronym,
+}
+
+def random_chars(count):
+    return ''.join(random.choice(string.ascii_lowercase) for _ in range(count))
+
+def gen_id(row):
+    state = row['state']
+    state_abbrev = us_state_to_abbrev[state].lower()
+    city = row['city']
+    city_abbrev = city.replace(' ', '').replace('.', '').lower()
+    if state_abbrev == unknown_location_acronym:
+        city_abbrev = unknown_location_acronym
+    if state_abbrev == 'dc':
+        city_abbrev = 'dc'
+
+    # id_line = f'id: {state_abbrev}-{city_abbrev}-{city_index}'
+    return f"{state_abbrev}-{city_abbrev}-{random_chars(4)}"
+
+def rewrite_data(data):
+    state_to_rows = {}
+    for row in data:
+        state = row["state"]
+        if state not in state_to_rows:
+            state_to_rows[state] = []
+        state_to_rows[state].append(row)
+    
+    # We don't need to sort because we read these in the order of the markdown files
+    #for state, data_rows in state_to_rows.items():
+    #    data_rows.sort(key=lambda row: row["city"] + row["id"])
+
+    for state, data_rows in state_to_rows.items():
+        out_path = f"{data_builder.md_dir}/{state}.md"
+        #print(out_path)
+        # links, city, description, name, date_text, id
+        new_md_text = gen_md_from_rows(state, data_rows)
+        with open(out_path, "wb") as fout:
+            fout.write(new_md_text.encode("utf-8"))
+
+row_format = '''### {name} | {date_text}
+
+{description}
+
+tags: {tags_md}
+
+id: {id}
+
+**Links**
+
+{links_md}
+
+'''
+
+def markdown_link(link_obj):
+    url = link_obj["url"]
+    text = link_obj["text"]
+    if len(text) == 0:
+        return url
+    return f"[{text}]({url})"
+
+def gen_md_from_rows(state, rows):
+    city = ''
+    lines = []
+    for row in rows:
+        if row["city"] and row["city"] != city:
+            # new city, let everyone know
+            lines.append(f'## {row["city"]}\n')
+            city = row["city"]

```

Comment:
this logic could be broken by incorrect manual additions, but I suppose we can just mitigate that with careful reviews

Reply:
Yeah maybe a typo on the city could do that, but I think it would be clear enough in the PR diff that we could find it

Reply:
Yes. There even is a current single degenerate case in the data. `https://www.instagram.com/p/CBEr2emJRyJ/?igshid=oloj7lgtv81t (video on third slide)` considers the entire text to be one big link. I plan to fix that in a different PR.

Pretty much every logic we build to parse this can be broken by bad data entry. Putting the regeneration in the PR will make that visible instead of silently broken.